### PR TITLE
Fix hard armor damage without breaking ZL

### PR DESCRIPTION
**Pull request by perkinslr**
## Changes

Go back to the hard armor damage formula that properly handles near-pen deflected damage properly.

## References

Links to the associated issues or other related pull requests, e.g.
- Partially reverts #1437's revert from #1193.
- Related to #1437 and #1504

## Reasoning

Why did you choose to implement things this way, e.g.
- Hard armor needs to take damage from attacks that *almost* pen.
- Clamping 0/1 should avoid the ZL issue

## Alternatives

Square or sqrt the relative AP values to tweak damage taken.
Don't cap the top end damage - probably break ZL again, but maybe use a higher cap?

## Testing

Check tests you have performed:
- [X] Compiles without warnings
- [X] Game runs without errors
- [ ] (For compatibility patches) ...with and without patched mod loaded
- [ ] Playtested a colony (specify how long)

### CODE REVIEW

**Comment by AluminumAlman** [comment]
So this makes it so that armor takes the absorbed amount of damage, multiplied by the penetration to armor ratio; wouldn't that cause better quality armor to take more damage? How I imagine it, awful steel plate armor would last longer than a legendary one against weapons that have more penetration than the armor value of the legendary plate armor because awful steel plate armor just absorbs less damage.

**Comment by AluminumAlman** [comment]
With these changes to hard armor damage, perhaps the PR could also remove https://github.com/CombatExtended-Continued/CombatExtended/blob/c3d18688a61457d2dd5a19bcfd16f4223ed61327/Source/CombatExtended/CombatExtended/ArmorUtilityCE.cs#L313-L317? I believe it isn't necessary with the new formula.

**Comment by AluminumAlman** [comment]
Did 8 tests with normal steel plate armor, normal steel longsword, level 8 melee traitless pawn against 100 squirrels, and my results were 40.625% average, 44% median, 20% min, 55% max durability with the change.

**Comment by ItsSamGray** [comment]
> Did 8 tests with normal steel plate armor, normal steel longsword, level 8 melee traitless pawn against 100 squirrels, and my results were 40.625% average, 44% median, 20% min, 55% max durability with the change.

So, are we back to squirrels damaging hard armor?

**Comment by AluminumAlman** [comment]
> So, are we back to squirrels damaging hard armor?

Huh, I thought it still did, but in 3.6 it doesn't (tested two times with normal steel plate armor: no damage). Well, either way, I believe that the change is good overall.

**Comment by perkinslr** [comment]
While I don't mind a squirrel doing a *little* damage to plate armor once in a while, and bearing in mind the damage done to armor will be increased by having partial armor enabled in the settings, I do think 55% max durability is too low.  
Plate armor has 1.8 <-> 3 mmRHA and 2.7-4.5 MPa, vs 0.07mm AP and 0.675MPa for the squirrel's bite, at 4 damage.  
If we square the first term, that drops the expected damage from 0.155 to 0.006 from the sharp damage, which I think would be fine.  Expected damage is 1 after 200 bites.  The blunt damage is still a problem.  Squaring the first term reduces the expected damage from 1 per bite to 0.25 per bite, but I think that's still too high.  

I think we need to differentiate sharp and blunt damage for hard armor.  While a sharp attack by something near a metal's hardness will at least leave scratches, and those scratches will add up over time, blunt impact deforming a material less than its elastic point will generally do nothing (in fine point, it may cause some localized hardening or softening of the material, but that effect is too minor to bother tracking).  

Bottom line, I think there should be a threshold, somewhere around half the armor's MPa rating, below which it takes no damage.

almansquared

**Comment by AluminumAlman** [comment]
95% average durability left with the new changes; 70% average durability left without squaring; 88% durability left when `if (isSharpDmg && (penAmount / armorAmount) < 0.5f)` is changed to ` if (!isSharpDmg && (penAmount < armorAmount * 0.5f))`.

**Review by N7Huntsman — DISMISSED** [review]
Everything looks normal during testing.

**Review by trikanyo — APPROVED** [review]

**Review by trikanyo — APPROVED** [review]

**Review by N7Huntsman — APPROVED** [review]
Appears all good.
